drivers/counter/counter_native_posix.c

```diff
@@ -80,41 +124,88 @@ static uint32_t ctr_get_pending_int(const struct device *dev)
 	return 0;
 }
 
+static bool is_any_alarm_pending(void)
+{
+	for (int i = 0; i < DRIVER_CONFIG_INFO_CHANNELS; i++) {
+		if (is_alarm_pending[i]) {
+			return true;
+		}
+	}
+	return false;
+}
+
 static int ctr_set_top_value(const struct device *dev,
 			     const struct counter_top_cfg *cfg)
 {
 	ARG_UNUSED(dev);
-	ARG_UNUSED(cfg);
 
-	posix_print_warning("%s not supported\n", __func__);
-	return -ENOTSUP;
+	if (is_any_alarm_pending()) {
+		posix_print_warning("Can't set top value while alarm is active\n");
+		return -EBUSY;
+	}
+
+	uint32_t current_value = hw_counter_get_value();
+
+	if (cfg->flags & COUNTER_TOP_CFG_DONT_RESET) {
+		if (current_value >= cfg->ticks) {
+			if (cfg->flags & COUNTER_TOP_CFG_RESET_WHEN_LATE) {
+				hw_counter_reset();
+			}
+			return -ETIME;
+		}
+	} else {
+		hw_counter_reset();
+	}
+
+	top = *cfg;
+	hw_counter_set_wrap_value((uint64_t)top.ticks + 1);
+
+	if ((cfg->ticks == TOP_VALUE) && !cfg->callback) {
+		is_top_set = false;
+	} else {
+		is_top_set = true;
+	}
+
+	schedule_next_isr();
+
+	return 0;
 }
 
 static uint32_t ctr_get_top_value(const struct device *dev)
 {
-	return TOP_VALUE;
+	return top.ticks;
 }
 
 static int ctr_set_alarm(const struct device *dev, uint8_t chan_id,
 			 const struct counter_alarm_cfg *alarm_cfg)
 {
 	ARG_UNUSED(dev);
 
-	if (chan_id >= DRIVER_CONFIG_INFO_CHANNELS) {
-		posix_print_warning("channel %u is not supported\n", chan_id);
-		return -ENOTSUP;
-	}
+	if (is_alarm_pending[chan_id])
+		return -EBUSY;
 
-	pending_alarm = *alarm_cfg;
-	is_alarm_pending = true;
+	uint32_t ticks = alarm_cfg->ticks;
+
+	if (ticks > top.ticks) {
+		posix_print_warning("Alarm ticks %u exceed top ticks %u\n", ticks,
+				top.ticks);
+		return -EINVAL;
+	}
 
 	if (!(alarm_cfg->flags & COUNTER_ALARM_CFG_ABSOLUTE)) {
-		pending_alarm.ticks =
-			hw_counter_get_value() + pending_alarm.ticks;
+		uint32_t current_value = hw_counter_get_value();
+
+		ticks += current_value;
+		if (ticks > top.ticks) { /* Handle wrap arounds */
+			ticks -= (top.ticks + 1); /* The count period is top.ticks + 1 */
+		}
 	}
 
-	hw_counter_set_target(pending_alarm.ticks);
-	irq_enable(COUNTER_EVENT_IRQ);
+	pending_alarm[chan_id] = *alarm_cfg;
+	pending_alarm[chan_id].ticks = ticks;
+	is_alarm_pending[chan_id] = true;
+
+	schedule_next_isr();
```

Comment:
Do we also need to call `hw_counter_set_wrap_value` here? Previous top configuration may have set it less than the alarm ticks.

Reply:
I'm not sure I follow. The wrap is configured at init, or every time ctr_set_top_value() is called successfully.
And the hw wrap remains in effect after that for ever.

Apart from that, we are checking in 189 and 199 that the alarm target does not go beyond the top.tick / wrap_value.